### ISSUE TITLE
Sequence Format v1

### ISSUE THREAD

**Issue by ksen0** [tiny-tool-box/aseq#6]
Sequence in question is 15+ minutes altogether and contains the following basic blocks. (Capitalisation is not essential. Sanskit in parens. Asymmetric cues given as First/Second)

1. Classical Sun Salutations (Surya Namaskar)
2. Standing With (R/L), Inhale Knee up to Hip-Height
3. Palms Together (Anjali Mudra), Exhale Twist to (R/L)
4. Inhale Open Chest and Exhale (L/R) Hand to the Knee, and (R/L) Optionally to the Back, and Look over the (R/L) hand.
5. Triangle (Trikonasana) Step back with the (L/R) foot perpendicular to the front of the mat. The R/L foot points to the front of the mat. The hips face the long edge of the mat. With hands on the shoulders, or out to sides, hinge to the (R/L) and then, drop the R/L hand down. The L/R arm can reach for the ceiling, or be tucked behind your back, to help keep the chest open.
6. Revolved Triangle (Parivrtta Trikonasana)
7. Practice W1->W3 (Virabhadrasana I -> Virabhandrasana III) transition. Tadāsana -> Inhale lift R/L foot -> Exhale step to Warrior 1 -> Inhale palms together and up -> Exhale hands down and interlaced behind the back for Humble Warrior -> Bring the R/L heel up and transfer weight to the L/R foot, optionally lifting the R/L foot and straightening the L/R leg for Warrior 3.
8. After W1->W3, add Revolved Half-Moon (Parivrtta Ardha Chandrasana). Same as above, but after W3, look down, drop the R/L hand to the floor or a block, and try to use the exhale to revolve to the left. The hips stay facing the mat, and the L/R hand can optionally come behind you or reach up, after you have already found the twist.

Then the flow/timing is:
* 1 for 6 minutes (about 1 minute per round per side, so even number of minutes, with clear indication when 1 minute is left)
* 2->3->4 on Right and Left side, 2 times (3 breaths on each side, so at around 8-10 seconds/breath, that is 3*2*2*10/60) ~ 2 minutes
* 5 for 45 seconds on each side (1.5 total)
* 6 for 45 seconds on each side (1.5 total)
* 7 for 30 seconds on each side, 2 times (2 minutes)
* 8 for 1 minute on each side (2 minutes)

The idea is that you're welcome to keep playing with the last pose if you feel like it, but everything up to it prepares you for it.

**Comment by ericHgorski** [comment]
Pinned question: how much information should be displayed on each card. Title with smaller description? Or all same size?

**Comment by ksen0** [comment]
7. Warrior I --> Warrior III Transition
8. Revolved Half-Moon (Parivrtta Ardha Chandrasana)

**Comment by ksen0** [comment]
@ericHgorski Regarding multi-card sections in the sequence:

In `https://github.com/ksen0/aseq/blob/master/posesAndSequence.js` the `sequence` can rather be a nested list of lists:

`const sequence = [
 [ {
        pose: poses[0],
        duration: 2,
    }],
[
    {
        pose: poses[1],
        duration: 3,
    },
    {
        pose: poses[2],
        duration: 1,
    },
    {
        pose: poses[3],
        duration: 4,
    },
]
]`

For each element in the sequence, `switchSide` is true if ANY of the poses have `switchside` true.

So you can iterate through each pose, in each list; so [this loop](https://github.com/ksen0/aseq/blob/master/script.js#L18) becomes nested.

**Comment by ericHgorski** [comment]
> @ericHgorski Regarding multi-card sections in the sequence:
> 
> In `https://github.com/ksen0/aseq/blob/master/posesAndSequence.js` the `sequence` can rather be a nested list of lists:
> 
> `const sequence = [ [ { pose: poses[0], duration: 2, }], [ { pose: poses[1], duration: 3, }, { pose: poses[2], duration: 1, }, { pose: poses[3], duration: 4, }, ] ]`
> 
> For each element in the sequence, `switchSide` is true if ANY of the poses have `switchside` true.
> 
> So you can iterate through each pose, in each list; so [this loop](https://github.com/ksen0/aseq/blob/master/script.js#L18) becomes nested.

Still trying to wrap my head around this! Ive implemented the nested arrays structure you suggested and adapted the loops to work for them. But Im still a bit stuck in terms of getting the desired sequencing behavior. I can think of a convoluted solution that checks if each item in the sequence array is an array of length > 1 and then write a custom if block to get what we want... but it felt like you maybe had a more elegant approach? What do you reccomend?

Not sure why this whole array and sequencing business is posing such a problem for me! My apologies.

**Comment by ksen0** [comment]
@ericHgorski no worries, so I looked at your implementation and here's an idea. What if when you generate the cards, you also generate an array `other_side_start` such that `other_side_start[postIndex]` is exactly the index of the pose where the other side should start, OR -1/undefined? so in our case:

`other_side_start=[0, -1, -1, 1, 4, 5, 6, 7]`

So, when global var is generated in the nested loop `$.each(sequence, (i, info) => {` all poses in that loop get -1 except the last one, which gets the start of the subsequence (only if at least one pose requires side switch!). This means you set -1 in each iteration of `info.forEach((item) => {` but then have a separate check on whether to set the last one to anything. In our example the last is set to 1.

Then instead of `if (currentPose.data("switchside")) {` on line 50 you can check IF `other_side_start[postIndex]>=0` and then activate that card (not by default the current card)

WDYT?

**Comment by ericHgorski** [comment]
I got the array to generate properly, but when I tried to use it in the modification to line 50, I got behavior where it just got stuck on a single card. I think part of the problem for me is that Im not 100% on exactly what the intended behavior is? Im pretty confident I do but perhaps it would be worth having a call to chat about it? We could also talk briefly about the layout stuff, as I believe that is the last outstanding issue!
